File: routes/wetter.py
import tempfile
from pathlib import Path

from flask import Blueprint, current_app, jsonify, request
from werkzeug.utils import secure_filename

from core.wetter import einlesen, openmeteo, speicher

bp = Blueprint("wetter", __name__, url_prefix="/api/wetter")


@bp.get("")
def liste():
    return jsonify(speicher.datensaetze())


@bp.patch("/<int:datensatz_id>")
def umbenennen(datensatz_id):
    daten = request.get_json(force=True) or {}
    if not daten.get("name"):
        return jsonify({"fehler": "Feld 'name' fehlt oder ist leer"}), 400
    try:
        speicher.datensatz_umbenennen(datensatz_id, daten["name"])
    except KeyError as fehler:
        return jsonify({"fehler": str(fehler)}), 404
    return jsonify({"ok": True})


@bp.delete("/<int:datensatz_id>")
def loeschen(datensatz_id):
    try:
        speicher.datensatz_loeschen(datensatz_id)
    except ValueError as fehler:
        return jsonify({"fehler": str(fehler)}), 400
    return jsonify({"ok": True})


@bp.post("/abrufen")
def abrufen():
    """Ruft ein Jahr Wetterdaten fuer einen Ort ueber die Open-Meteo Archive-API
    ab und legt sie genauso ab wie ein Datei-Upload. Dauert - bei fuenf parallel
    geschickten Teilabfragen - in aller Regel nur wenige Sekunden; ein eigener
    Hintergrundlauf mit Fortschrittsanzeige (wie core.laeufe fuer Simulationen)
    waere hier ueberdimensioniert, siehe wetter-api-report.md.
    """
    daten = request.get_json(silent=True) or {}

    try:
        breite = float(daten["breite"])
        laenge = float(daten["laenge"])
        jahr = int(daten["jahr"])
    except (KeyError, TypeError, ValueError):
        return jsonify({
            "fehler": "Breite, Länge und Jahr werden benötigt (Breite/Länge als "
                      "Dezimalzahl, Jahr als Ganzzahl)"
        }), 400

    ort = (daten.get("ort") or "").strip()
    name = (daten.get("name") or "").strip() or f"{ort or f'{breite}, {laenge}'} {jahr}"

    try:
        stunden = openmeteo.abrufen(breite, laenge, jahr, ort=ort)
    except openmeteo.WetterEingabeFehler as fehler:
        return jsonify({"fehler": str(fehler)}), 400
    except openmeteo.WetterAbrufFehler as fehler:
        current_app.logger.warning("Wetterabruf gescheitert: %s", fehler)
        return jsonify({"fehler": str(fehler)}), 502

    datensatz_id = speicher.datensatz_anlegen(
        name, "open-meteo", stunden, ort=ort, breite=breite, laenge=laenge, jahr=jahr,
        notiz=f"Abgerufen von Open-Meteo (Archive-API) für "
              f"{ort or f'{breite}, {laenge}'}, Jahr {jahr}",
    )
    return jsonify({"id": datensatz_id, "stunden": len(stunden)}), 201


@bp.post("/upload")
def hochladen():
    """Nimmt eine oder mehrere Wetterdateien an und legt je Datei einen
    Datensatz an.

    Der uebliche Weg ist ein ganzer Ordner: der Browser sammelt daraus die
    lesbaren Dateien und schickt nur die angekreuzten (static/js/start.js).
    Ein Einzelupload ist der Sonderfall mit genau einer Datei; damit aeltere
    Aufrufer weiterlaufen, traegt die Antwort dann zusaetzlich 'id' und
    'stunden' unmittelbar.

    'fehler' ist wie ueberall sonst in dieser Schnittstelle eine Meldung als
    Text und nur bei einem Fehlschlag da; was einzelne Dateien betrifft, steht
    unter 'dateifehler'.

    Jede Datei wird fuer sich verarbeitet. Eine unlesbare darf die uebrigen
    nicht mitreissen - wer sechs Testreferenzjahre hochlaedt und bei einem
    einen Uebertragungsfehler hat, soll die anderen fuenf behalten und
    erfahren, welches gefehlt hat.
    """
    dateien = [d for d in request.files.getlist("datei") if d and d.filename]
    if not dateien:
        return jsonify({"fehler": "Keine Datei übermittelt"}), 400

    ort = (request.form.get("ort") or "").strip()
    # Ein mitgegebener Name kann nur fuer eine einzelne Datei gelten - sonst
    # hiessen alle Datensaetze gleich.
    vorgegebener_name = (request.form.get("name") or "").strip() if len(dateien) == 1 else ""

    angelegt = []
    fehler = []
    for datei in dateien:
        try:
            angelegt.append(_datei_uebernehmen(datei, ort, vorgegebener_name))
        except ValueError as ausnahme:
            fehler.append({"datei": datei.filename, "fehler": str(ausnahme)})

    if not angelegt:
        # Bei einer einzelnen Datei ist deren Meldung die ganze Wahrheit; bei
        # mehreren waere eine Aneinanderreihung unlesbar.
        meldung = (
            fehler[0]["fehler"] if len(fehler) == 1
            else f"Keine der {len(fehler)} Dateien ließ sich lesen."
        )
        return jsonify({"fehler": meldung, "datensaetze": [], "dateifehler": fehler}), 400

    antwort = {"datensaetze": angelegt, "dateifehler": fehler}
    if len(angelegt) == 1:
        antwort["id"] = angelegt[0]["id"]
        antwort["stunden"] = angelegt[0]["stunden"]
    return jsonify(antwort), 201


def _datei_uebernehmen(datei, ort, vorgegebener_name):
    """Liest eine hochgeladene Datei und legt ihren Datensatz an.

    Wirft ValueError mit einer verstaendlichen deutschen Meldung, wenn die
    Datei nicht lesbar ist - der Aufrufer sammelt diese Meldungen je Datei.
    """
    endung = Path(datei.filename).suffix.lower()
    if endung in einlesen.ARCHIVENDUNGEN:
        raise ValueError(
            "Archive werden nicht ausgepackt. Bitte das Archiv entpacken und "
            "den entpackten Ordner auswählen – die Dateien darin werden dann "
            "gefunden, auch in Unterordnern."
        )
    if endung not in einlesen.ENDUNGEN:
        benannt = endung or "ohne Endung"
        raise ValueError(
            f"'{benannt}' wird nicht gelesen. Erwartet werden "
            f"{', '.join(einlesen.ENDUNGEN)}."
        )

    # Unter ihrem urspruenglichen Namen ablegen, nicht unter einem zufaelligen:
    # ein TRY traegt Schluesseljahr und Art im Dateinamen (Handbuch Kap. 2), und
    # der Leser braucht das Jahr - die Datenzeilen selbst nennen keines. Mit
    # einem Namen wie "tmp8f3k.dat" bekaeme jedes Zukunfts-TRY stillschweigend
    # das Vorgabejahr 2015.
    with tempfile.TemporaryDirectory() as ordner:
        pfad = Path(ordner) / _sicherer_dateiname(datei.filename, endung)
        datei.save(str(pfad))
        try:
            stunden = einlesen.lese_datei(pfad)
            kopf = einlesen.beschreibung(pfad)
        except ValueError:
            # Kommt schon mit einer verstaendlichen Meldung - unveraendert weiter.
            raise
        except Exception as ausnahme:
            # Beschaedigte oder falsch benannte Dateien melden je nach Bibliothek sehr
            # verschiedene Fehler - xlrd, openpyxl und das Auspacken des Zip-Behaelters
            # haben nichts gemeinsam. Wer eine kaputte Datei hochlaedt, soll einen Satz
            # lesen und keine Fehlerseite.
            current_app.logger.exception("Wetterdatei nicht lesbar: %s", datei.filename)
            raise ValueError(
                "Die Datei ließ sich nicht lesen. Erwartet wird ein DWD-Testreferenzjahr "
                "(.dat) oder eine Tabelle im Format des Blattes 'Wetterdaten'."
            ) from ausnahme

    if not stunden:
        raise ValueError("Die Datei enthält keine Stundenwerte")

    name = vorgegebener_name or _name_finden(datei.filename, kopf)
    datensatz_id = speicher.datensatz_anlegen(
        name, "upload", stunden, ort=ort,
        breite=kopf.get("breite"), laenge=kopf.get("laenge"),
        jahr=kopf.get("jahr"), notiz=kopf.get("notiz", ""),
    )
    return {
        "datei": datei.filename,
        "id": datensatz_id,
        "name": name,
        "stunden": len(stunden),
    }


def _sicherer_dateiname(dateiname, endung):
    """Der Dateiname kommt vom Browser und darf nicht ungeprueft in einen Pfad.

    secure_filename streicht Pfadanteile und Sonderzeichen; bleibt dabei nichts
    Brauchbares uebrig (etwa bei einem rein kyrillischen Namen), tritt ein
    Ersatzname ein - dann fehlt zwar die Kennung aus dem Namen, aber die Datei
    laesst sich trotzdem lesen.
    """
    name = secure_filename(Path(dateiname).name)
    if not name:
        name = f"wetterdatei{endung}"
    if not name.lower().endswith(endung):
        name += endung
    return name


def _name_finden(dateiname, kopf):
    """Ein TRY traegt Jahr und Art im Namen ('2015 – mittleres Jahr'); der Ort
    steht am Datensatz und wuerde hier nur doppelt stehen. Alles andere behaelt
    seinen Dateinamen."""
    if kopf.get("jahr") and kopf.get("art"):
        return f"{kopf['jahr']} – {kopf['art']}"
    return Path(dateiname).stem

File: static/js/wetter.js
/* Wetterseite: die Datensaetze verwalten und neue holen (Online-Abruf oder
   eigene TRY-Dateien).

   Die Liste selbst steht fertig in der Seite (templates/wetter.html, gefuellt
   in routes/pages.py: wetter()) - dieses Modul verdrahtet nur Umbenennen und
   Loeschen daran und bedient die beiden Formulare zum Hinzufuegen.

   Bis zum Umbau der Oberflaeche steckte all das im gemeinsamen start.js, das
   beide Seiten in voller Laenge luden: Die Wetterseite bekam die
   Projektdialoge mitgeliefert, die Startseite die Ordnerauswertung - zusammen
   44 kB, von denen jede Seite gut die Haelfte nie anfasste. zeigeFehler(),
   htmlSicher() und die beiden Dialoge stehen in static/js/dialoge.js. */

// Erstes Jahr, das die Open-Meteo Archive-API anbietet (core/wetter/openmeteo.py:
// FRUEHESTES_JAHR) - hier dupliziert, weil die Grenze rein in der Oberflaeche
// gebraucht wird und dafuer keine eigene Schnittstelle lohnt.
const WETTER_FRUEHESTES_JAHR = 1940;

// Endungen, die der Server lesen kann (core/wetter/einlesen.py: ENDUNGEN).
// Wonach hier nicht gefiltert wird, geht auch nicht ueber die Leitung - der
// Ordner eines Testreferenzjahrs enthaelt neben den Daten ein Handbuch von
// rund zwei Megabyte, das auf dem Rechner des Nutzers bleiben soll.
const WETTER_ENDUNGEN = [".dat", ".xls", ".xlsx", ".xlsm", ".csv"];

// Art des TRY aus dem Kuerzel im Dateinamen (Handbuch Kap. 2, AAAA) - dieselbe
// Zuordnung wie in core/wetter/try_dat.py: ARTEN. Hier nur fuer die
// Beschriftung der Ankreuzliste; verbindlich benannt wird auf dem Server.
const TRY_ARTEN = {
  jahr: "mittleres Jahr",
  somm: "extremer Sommer",
  wint: "extremer Winter",
};
const TRY_NAMENSMUSTER = /^TR[YJ](\d{4})_[^_]*_(Jahr|Somm|Wint)/i;


// Archive werden bewusst nicht ausgepackt (core/wetter/einlesen.py:
// ARCHIVENDUNGEN). Sie werden hier trotzdem erkannt, damit ein Nutzer, der ein
// heruntergeladenes Zip waehlt, erfaehrt, was er tun soll - statt vor einer
// leeren Liste zu stehen.
const WETTER_ARCHIVENDUNGEN = [".zip", ".7z", ".rar", ".tar", ".gz", ".tgz", ".bz2"];

const ARCHIV_HINWEIS =
  "Archive werden nicht ausgepackt. Bitte das Archiv entpacken und den " +
  "entpackten Ordner auswählen – die Dateien darin werden dann gefunden, " +
  "auch in Unterordnern.";

function wetterIstArchiv(dateiname) {
  const klein = dateiname.toLowerCase();
  return WETTER_ARCHIVENDUNGEN.some((endung) => klein.endsWith(endung));
}

function wetterEndungErlaubt(dateiname) {
  const klein = dateiname.toLowerCase();
  return WETTER_ENDUNGEN.some((endung) => klein.endsWith(endung));
}

// "TRY2015_510881137633_Somm.dat" -> "2015 - extremer Sommer". Alles, was nicht
// auf die Konvention passt, behaelt seinen Dateinamen: raten waere schlechter
// als zeigen, was dasteht.
function wetterBeschriftung(dateiname) {
  const treffer = TRY_NAMENSMUSTER.exec(dateiname);
  if (!treffer) return dateiname;
  const art = TRY_ARTEN[treffer[2].toLowerCase()];
  return art ? `${treffer[1]} – ${art}` : dateiname;
}

// Schluesseljahr aus dem Dateinamen, oder null.
function wetterSchluesseljahr(dateiname) {
  const treffer = TRY_NAMENSMUSTER.exec(dateiname);
  return treffer ? Number(treffer[1]) : null;
}

// Ein Zukunfts-TRJ beruht auf Klimamodellen fuer einen Zeitraum, der noch
// bevorsteht (derzeit 2031-2060, Schluesseljahr 2045). Es wird seltener
// gebraucht als das Gegenwarts-TRJ und darum nicht vorgehakt - man soll es
// bewusst dazunehmen.
//
// Verglichen wird mit dem laufenden Jahr statt mit der festen Zahl 2045:
// liefert der DWD spaeter einen weiteren Zukunftsdatensatz unter einem anderen
// Schluesseljahr, greift die Regel weiterhin.
function wetterIstZukunftsjahr(dateiname) {
  const jahr = wetterSchluesseljahr(dateiname);
  return jahr !== null && jahr > new Date().getFullYear();
}

// Der oberste Ordnername der Auswahl - beim Ordner-Upload der Standort.
// webkitRelativePath sieht bei einer Ordnerwahl so aus:
// "Dresden_Industriegelaende/TRY_510881137633/TRY2015_..._Jahr.dat".
function wetterStandortRaten(dateien) {
  for (const datei of dateien) {
    const pfad = datei.webkitRelativePath || "";
    const teile = pfad.split("/").filter(Boolean);
    if (teile.length > 1) return teile[0];
  }
  return "";
}

const Wetter = {
  // Die im Ordner gefundenen und im Formular angekreuzten Dateien.
  wetterAuswahl: [],
  // Mitgewaehlte Archive - nur, um sie erklaeren zu koennen.
  wetterArchive: [],

  /* Die Wetterliste steht schon im HTML (templates/wetter.html, gefuellt in
     routes/pages.py: wetter()) - dieses Skript baut sie nicht mehr, es haengt
     nur seine Handlungen an die fertigen Zeilen. Vorher holte es sie nach dem
     Laden ueber /api/wetter nach: bis die Antwort da war, zeigte die Seite,
     deren einziger Inhalt diese Liste ist, eine leere Flaeche - und auf keinem
     Abzug der Seite war sie zu sehen.

     Nach einer Aenderung wird die Seite neu geladen, wie bei der Projektliste
     der Startseite (bindeProjekte). */
  bindeWetter() {
    document.querySelectorAll("#start-wetter-liste tbody tr[data-id]").forEach((zeile) => {
      const datensatz = {
        id: Number(zeile.dataset.id),
        name: zeile.dataset.name,
        stunden: Number(zeile.dataset.stunden),
      };
      zeile.querySelector(".wetter-umbenennen").addEventListener("click", () =>
        this.wetterUmbenennenDialog(datensatz)
      );
      zeile.querySelector(".wetter-loeschen").addEventListener("click", () =>
        this.wetterLoeschenDialog(datensatz)
      );
    });
  },

  async wetterUmbenennenDialog(datensatz) {
    const neuerName = await textEingabeDialog("Wetterdatensatz umbenennen", datensatz.name);
    if (neuerName === null) return;
    let antwort;
    try {
      antwort = await fetch(`/api/wetter/${datensatz.id}`, {
        method: "PATCH",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ name: neuerName }),
      });
    } catch {
      zeigeFehler("Wetterdatensatz konnte nicht umbenannt werden.");
      return;
    }
    if (!antwort.ok) {
      zeigeFehler("Wetterdatensatz konnte nicht umbenannt werden.");
      return;
    }
    await this._wetterListeAktualisieren(
      "Umbenannt, die Liste konnte aber nicht aktualisiert werden."
    );
  },

  // Der Loeschknopf ist bei einem verwendeten Datensatz schon deaktiviert
  // (templates/wetter.html) - dieser Zweig greift nur bei einem inzwischen
  // veralteten Stand (z.B. ein zweiter, offener Tab hat gerade einen Lauf
  // gestartet), nicht als erster Weg dorthin.
  async wetterLoeschenDialog(datensatz) {
    const bestaetigt = await bestaetigenDialog(
      "Wetterdatensatz löschen",
      `Wetterdatensatz "${htmlSicher(datensatz.name)}" (${datensatz.stunden} Stunden) ` +
        "wirklich löschen?"
    );
    if (!bestaetigt) return;
    let antwort;
    try {
      antwort = await fetch(`/api/wetter/${datensatz.id}`, { method: "DELETE" });
    } catch {
      zeigeFehler("Wetterdatensatz konnte nicht gelöscht werden.");
      return;
    }
    if (!antwort.ok) {
      let text = "Wetterdatensatz konnte nicht gelöscht werden.";
      try {
        const daten = await antwort.json();
        if (daten.fehler) text = daten.fehler;
      } catch {
        /* Antwort war kein JSON - bei der Vorgabemeldung bleiben. */
      }
      zeigeFehler(text);
      return;
    }
    await this._wetterListeAktualisieren(
      "Gelöscht, die Liste konnte aber nicht aktualisiert werden."
    );
  },

  // Sammelt die Auswahl aus beiden Dateifeldern (Ordner und Einzeldateien),
  // wirft alles Unlesbare weg und baut daraus die Ankreuzliste.
  //
  // Gefiltert wird hier und nicht erst auf dem Server, damit das Handbuch-PDF
  // aus einem TRY-Ordner den Rechner gar nicht erst verlaesst.
  wetterDateiAktualisieren() {
    const ordnerFeld = document.getElementById("feld-wetter-ordner");
    const dateiFeld = document.getElementById("feld-wetter-datei");
    const ortFeld = document.getElementById("feld-wetter-ort");

    const ordnerDateien = Array.from(ordnerFeld?.files || []);
    const einzelDateien = Array.from(dateiFeld?.files || []);
    const alle = [...ordnerDateien, ...einzelDateien];
    const lesbare = alle.filter((d) => wetterEndungErlaubt(d.name));
    this.wetterArchive = alle.filter((d) => wetterIstArchiv(d.name));

    this.wetterAuswahl = lesbare.map((datei) => ({
      datei,
      pfad: datei.webkitRelativePath || datei.name,
      beschriftung: wetterBeschriftung(datei.name),
      angehakt: !wetterIstZukunftsjahr(datei.name),
    }));

    const ordnerName = wetterStandortRaten(ordnerDateien);
    if (ordnerName && ortFeld && !ortFeld.value.trim()) ortFeld.value = ordnerName;

    const ordnerAnzeige = document.getElementById("feld-wetter-ordnername");
    if (ordnerAnzeige) {
      ordnerAnzeige.textContent = ordnerName
        ? `${ordnerName} (${ordnerDateien.length} Dateien)`
        : "Kein Ordner ausgewählt";
    }
    const dateiAnzeige = document.getElementById("feld-wetter-dateiname");
    if (dateiAnzeige) {
      dateiAnzeige.textContent =
        einzelDateien.length === 0
          ? "Keine Datei ausgewählt"
          : einzelDateien.map((d) => d.name).join(", ");
    }

    this.zeichneWetterAuswahl(alle.length - lesbare.length);
  },

  // Die Ankreuzliste unter den Dateifeldern. Sie zeigt auch, wie viele Dateien
  // aussortiert wurden - sonst wirkt ein Ordner mit Handbuch, als haette das
  // Programm etwas verschluckt.
  zeichneWetterAuswahl(uebersprungen) {
    const bereich = document.getElementById("wetter-auswahl-liste");
    if (!bereich) return;
    bereich.textContent = "";

    // Ein Archiv erklaeren wir auch dann, wenn sonst nichts gefunden wurde -
    // sonst bliebe der Kasten leer und der Nutzer ratlos.
    if (!this.wetterAuswahl.length && !this.wetterArchive.length) {
      bereich.hidden = true;
      return;
    }
    bereich.hidden = false;

    if (this.wetterArchive.length) {
      const archivHinweis = document.createElement("p");
      archivHinweis.className = "wetter-auswahl-hinweis wetter-auswahl-warnung";
      archivHinweis.textContent = ARCHIV_HINWEIS;
      bereich.appendChild(archivHinweis);
    }

    const liste = document.createElement("ul");
    liste.className = "wetter-auswahl-liste";
    this.wetterAuswahl.forEach((eintrag, nummer) => {
      const zeile = document.createElement("li");
      const marke = document.createElement("label");

      const haken = document.createElement("input");
      haken.type = "checkbox";
      haken.checked = eintrag.angehakt;
      haken.addEventListener("change", () => {
        this.wetterAuswahl[nummer].angehakt = haken.checked;
      });

      const titel = document.createElement("span");
      titel.className = "wetter-auswahl-titel";
      titel.textContent = eintrag.beschriftung;

      const pfad = document.createElement("span");
      pfad.className = "wetter-auswahl-pfad";
      pfad.textContent = eintrag.pfad;

      marke.append(haken, titel, pfad);
      zeile.appendChild(marke);
      liste.appendChild(zeile);
    });
    bereich.appendChild(liste);

    if (uebersprungen > 0) {
      const hinweis = document.createElement("p");
      hinweis.className = "wetter-auswahl-hinweis";
      hinweis.textContent =
        uebersprungen === 1
          ? "Eine weitere Datei im Ordner ist keine Wetterdatei und bleibt liegen."
          : `${uebersprungen} weitere Dateien im Ordner sind keine Wetterdaten und bleiben liegen.`;
      bereich.appendChild(hinweis);
    }
  },

  async wetterHochladen(ereignis) {
    ereignis.preventDefault();
    const ordnerFeld = document.getElementById("feld-wetter-ordner");
    const dateiFeld = document.getElementById("feld-wetter-datei");
    const ortFeld = document.getElementById("feld-wetter-ort");
    const nameFeld = document.getElementById("feld-wetter-name");
    const knopf = document.getElementById("btn-wetter-hochladen");

    const gewaehlt = this.wetterAuswahl.filter((e) => e.angehakt);
    if (!gewaehlt.length) {
      let meldung = "Bitte einen Ordner oder eine Datei auswählen.";
      if (this.wetterAuswahl.length) meldung = "Bitte mindestens eine Datei ankreuzen.";
      else if (this.wetterArchive.length) meldung = ARCHIV_HINWEIS;
      zeigeFehler(meldung);
      return;
    }

    const formular = new FormData();
    gewaehlt.forEach((eintrag) => formular.append("datei", eintrag.datei));
    if (ortFeld?.value.trim()) formular.append("ort", ortFeld.value.trim());
    // Eine eigene Bezeichnung kann nur fuer eine einzelne Datei gelten - bei
    // mehreren hiessen sonst alle Datensaetze gleich (routes/wetter.py).
    if (gewaehlt.length === 1 && nameFeld.value.trim()) {
      formular.append("name", nameFeld.value.trim());
    }

    const urspruenglicherText = knopf.textContent;
    knopf.disabled = true;
    knopf.textContent =
      gewaehlt.length === 1
        ? "Wird hochgeladen …"
        : `Wird hochgeladen (${gewaehlt.length}) …`;

    const zuruecksetzen = () => {
      knopf.disabled = false;
      knopf.textContent = urspruenglicherText;
    };

    let antwort;
    try {
      antwort = await fetch("/api/wetter/upload", { method: "POST", body: formular });
    } catch {
      zeigeFehler("Die Wetterdaten konnten nicht hochgeladen werden.");
      zuruecksetzen();
      return;
    }
    if (!antwort.ok) {
      let text = "Die Wetterdaten konnten nicht hochgeladen werden.";
      try {
        const daten = await antwort.json();
        if (daten.fehler) text = daten.fehler;
      } catch {
        /* Antwort war kein JSON - bei der Vorgabemeldung bleiben. */
      }
      zeigeFehler(text);
      zuruecksetzen();
      return;
    }

    // Ein Teilerfolg ist der Regelfall wert, gemeldet zu werden: wer sechs
    // Jahre hochlaedt und fuenf bekommt, soll nicht raten muessen, welches
    // fehlt.
    try {
      const daten = await antwort.json();
      if (daten.dateifehler?.length) {
        zeigeFehler(
          `Nicht gelesen: ${daten.dateifehler
            .map((f) => `${f.datei} (${f.fehler})`)
            .join("; ")}`
        );
      }
    } catch {
      /* Antwort war kein JSON - der Upload hat trotzdem geklappt. */
    }

    this.wetterArchive = [];
    if (ordnerFeld) ordnerFeld.value = "";
    if (dateiFeld) dateiFeld.value = "";
    if (ortFeld) ortFeld.value = "";
    nameFeld.value = "";
    this.wetterAuswahl = [];
    this.wetterDateiAktualisieren();
    zuruecksetzen();

    await this._wetterListeAktualisieren(
      "Die Daten wurden hochgeladen, die Liste konnte aber nicht aktualisiert werden."
    );
  },

  /* Nach einem Upload, Abruf, Umbenennen oder Loeschen die Seite neu laden,
     statt die Liste im Browser nachzuziehen: EINE Anfrage, und sie kann nicht
     von dem abweichen, was der Server ohnehin liefert. Die mitgegebene Meldung
     wird gebraucht, wenn schon das Neuladen scheitert - der Vorgang selbst ist
     dann durch, nur die Anzeige haengt hinterher. */
  async _wetterListeAktualisieren(fehlerBeiFehlschlag) {
    try {
      window.location.reload();
    } catch {
      zeigeFehler(fehlerBeiFehlschlag);
    }
  },

  // Fuellt das Jahr-Mehrfachauswahlfeld mit allen abgeschlossenen
  // Kalenderjahren, die die Open-Meteo Archive-API anbietet (1940 bis zum
  // Vorjahr) - neuestes zuerst, weil das der haeufigste Wunsch ist. So kann
  // die Oberflaeche gar nicht erst ein unzulaessiges Jahr anbieten, statt den
  // Benutzer erst beim Absenden auf den Fehler laufen zu lassen.
  wetterAbrufJahreFuellen() {
    const feld = document.getElementById("feld-wetter-abruf-jahre");
    if (!feld) return;
    const letztesVollstaendigesJahr = new Date().getFullYear() - 1;
    feld.textContent = "";
    for (let jahr = letztesVollstaendigesJahr; jahr >= WETTER_FRUEHESTES_JAHR; jahr--) {
      const option = document.createElement("option");
      option.value = String(jahr);
      option.textContent = String(jahr);
      feld.appendChild(option);
    }
    // Bequemer Einstieg: das juengste verfuegbare Jahr ist vorausgewaehlt.
    if (feld.options.length) feld.options[0].selected = true;
  },

  // Zeigt/versteckt die Koordinatenfelder, je nachdem ob ein vorbelegter Ort
  // oder "Eigene Koordinaten" gewaehlt ist.
  wetterAbrufOrtGewaehlt() {
    const auswahl = document.getElementById("feld-wetter-abruf-ort");
    const koordinatenBereich = document.getElementById("wetter-abruf-koordinaten");
    koordinatenBereich.hidden = auswahl.value !== "eigene";
  },

  async wetterAbrufen(ereignis) {
    ereignis.preventDefault();

    const ortAuswahl = document.getElementById("feld-wetter-abruf-ort");
    const jahreFeld = document.getElementById("feld-wetter-abruf-jahre");
    const nameFeld = document.getElementById("feld-wetter-abruf-name");
    const knopf = document.getElementById("btn-wetter-abrufen");

    let ort;
    let breite;
    let laenge;
    if (ortAuswahl.value === "eigene") {
      const breiteFeld = document.getElementById("feld-wetter-abruf-breite");
      const laengeFeld = document.getElementById("feld-wetter-abruf-laenge");
      const ortsnameFeld = document.getElementById("feld-wetter-abruf-ortsname");
      breite = parseFloat(breiteFeld.value);
      laenge = parseFloat(laengeFeld.value);
      if (!Number.isFinite(breite) || !Number.isFinite(laenge)) {
        zeigeFehler("Bitte Breite und Länge als Zahl eingeben.");
        return;
      }
      ort = ortsnameFeld.value.trim() || `${breite}, ${laenge}`;
    } else {
      const gewaehlteOption = ortAuswahl.selectedOptions[0];
      breite = parseFloat(gewaehlteOption.dataset.breite);
      laenge = parseFloat(gewaehlteOption.dataset.laenge);
      ort = ortAuswahl.value;
    }

    const jahre = Array.from(jahreFeld.selectedOptions).map((o) => parseInt(o.value, 10));
    if (!jahre.length) {
      zeigeFehler("Bitte mindestens ein Jahr auswählen.");
      return;
    }

    const eigenerName = nameFeld.value.trim();
    const urspruenglicherText = knopf.textContent;
    knopf.disabled = true;

    // Ein Abruf je Jahr, hintereinander statt parallel - der Server schickt
    // pro Jahr bereits fuenf parallele Teilabfragen an Open-Meteo los,
    // mehrere Jahre gleichzeitig wuerden das unnoetig vervielfachen. Fehler
    // bei einem Jahr sollen die uebrigen Jahre nicht verhindern.
    const fehlgeschlagen = [];
    let erfolge = 0;
    for (let i = 0; i < jahre.length; i++) {
      const jahr = jahre[i];
      knopf.textContent =
        jahre.length > 1 ? `Wird abgerufen … (${i + 1}/${jahre.length})` : "Wird abgerufen …";

      const name = eigenerName ? (jahre.length > 1 ? `${eigenerName} ${jahr}` : eigenerName) : "";

      let antwort;
      try {
        antwort = await fetch("/api/wetter/abrufen", {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ breite, laenge, jahr, ort, name }),
        });
      } catch {
        fehlgeschlagen.push(`${jahr}: Wetterdaten konnten nicht abgerufen werden`);
        continue;
      }
      if (!antwort.ok) {
        let text = "Wetterdaten konnten nicht abgerufen werden";
        try {
          const daten = await antwort.json();
          if (daten.fehler) text = daten.fehler;
        } catch {
          /* Antwort war kein JSON - bei der Vorgabemeldung bleiben. */
        }
        fehlgeschlagen.push(`${jahr}: ${text}`);
        continue;
      }
      erfolge++;
    }

    knopf.disabled = false;
    knopf.textContent = urspruenglicherText;

    if (fehlgeschlagen.length) {
      zeigeFehler(
        erfolge
          ? `${erfolge} von ${jahre.length} Jahren abgerufen. Fehlgeschlagen: ${fehlgeschlagen.join("; ")}`
          : `Abruf fehlgeschlagen: ${fehlgeschlagen.join("; ")}`
      );
    }

    if (erfolge) {
      nameFeld.value = "";
      await this._wetterListeAktualisieren(
        "Die Wetterdaten wurden abgerufen, die Liste konnte aber nicht aktualisiert werden."
      );
    }
  },
};

function verdrahte(kennung, ereignis, was) {
  const element = document.getElementById(kennung);
  if (element) element.addEventListener(ereignis, was);
}

window.addEventListener("DOMContentLoaded", () => {
  verdrahte("form-wetter-upload", "submit", (e) => Wetter.wetterHochladen(e));
  for (const kennung of ["feld-wetter-ordner", "feld-wetter-datei"]) {
    verdrahte(kennung, "change", () => Wetter.wetterDateiAktualisieren());
  }
  Wetter.wetterAbrufJahreFuellen();
  verdrahte("feld-wetter-abruf-ort", "change", () => Wetter.wetterAbrufOrtGewaehlt());
  verdrahte("form-wetter-abruf", "submit", (e) => Wetter.wetterAbrufen(e));

  // Die Wetterliste steht schon da - sie wird nur verdrahtet.
  Wetter.bindeWetter();
});
